### PR TITLE
Create pagination component

### PR DESCRIPTION
**Pull request by dhutaryan**
#11

### CODE REVIEW

File: src/app/heroes/pagination/pagination.component.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  ngOnInit() {
+    this.renderPaging();
+  }
+
+  renderPaging() {
+    const totalPages: number = Math.ceil(this.totalHeroes / this.pageSize);
+
+    this.pageList = new Array(totalPages)
+      .fill(0)
+      .map((item, i) => i + 1);
```
Comment (dbl7) [inline src/app/heroes/pagination/pagination.component.ts]: Try using full name `index`

File: src/app/heroes/pagination/pagination.component.html [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  <ul class="pagination">
+    <li class="page-item">
+      <a class="page-link" href="#" aria-label="Previous">
+        <span aria-hidden="true">&laquo;</span>
+        <span class="sr-only">Previous</span>
+      </a>
+    </li>
+    <li 
+      class="page-item"
+      *ngFor="let pageNo of pageList"
```
Comment (dbl7) [inline src/app/heroes/pagination/pagination.component.html]: What does it mean `pageNo`?!

File: src/app/heroes/pagination/pagination.component.html [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+      >
+        <a class="page-link" href="#">{{ pageNo }}</a>
```
Comment (dbl7) [inline src/app/heroes/pagination/pagination.component.html]: Don't use whitespaces around property name

File: src/app/heroes/pagination/pagination.component.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  }
```
Comment (dbl7) [inline src/app/heroes/pagination/pagination.component.ts]: Consider using Observable range for creating this pages

File: src/app/heroes/heroes.component.ts [hunk 5]
```diff
@@ -23,7 +24,8 @@ import { Hero } from '../models/hero';
 export class HeroesComponent implements OnInit {
   searchField: FormControl;
   heroes: Hero[];
-  loading: boolean = false;
+  count: number;
```
Comment (dbl7) [inline src/app/heroes/heroes.component.ts]: Rename count to `totalHeroes`

**Review by dbl7 — CHANGES_REQUESTED** [review]

File: src/app/heroes/heroes.component.ts (hunk 5 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.ts]: Fixed

File: src/app/heroes/pagination/pagination.component.ts (hunk 1 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.ts]: Fixed

File: src/app/heroes/pagination/pagination.component.ts (hunk 4 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.ts]: Done

File: src/app/heroes/pagination/pagination.component.html (hunk 3 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.html]: Fixed

File: src/app/heroes/pagination/pagination.component.html (hunk 2 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.html]: Changed to "pageNumber"

File: src/app/heroes/pagination/pagination.component.ts (hunk 4 above)
Comment (dbl7) [reply src/app/heroes/pagination/pagination.component.ts]: Use `Observable.range` for generating pages

**Review by dbl7 — APPROVED** [review]

File: src/app/heroes/pagination/pagination.component.ts (hunk 4 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.ts]: Fixed

File: src/app/heroes/heroes.component.html [hunk 6]
```diff
@@ -29,7 +29,11 @@ <h1>Star Wars <small>Encyclopedia</small></h1><hr/>
             </tbody>
           </table>
           <app-spinner *ngIf="loading$ | async"></app-spinner>
-          <app-pagination *ngIf="!(loading$ | async)" [totalHeroes]="totalHeroes"></app-pagination>
+          <app-pagination 
+            *ngIf="!(loading$ | async)"
```
Comment (dbl7) [inline src/app/heroes/heroes.component.html]: Get rid of this verification here

File: src/app/heroes/pagination/pagination.component.ts [hunk 7]
```diff
@@ -21,11 +24,14 @@ export class PaginationComponent implements OnInit {
   pagesCount() {
```
Comment (dbl7) [inline src/app/heroes/pagination/pagination.component.ts]: Rename this method to `generatePages()` or something like this

File: src/app/heroes/heroes.component.ts [hunk 8]
```diff
@@ -58,4 +62,13 @@ export class HeroesComponent implements OnInit {
       .do(() => this.loading$.next(false))
       .catch(err => Observable.throw(err));
   }
+
+  pageChanged(pageNumber: number) {
+    this.pageNumber$.next(pageNumber);
+  }
+
+  page() {
```
Comment (dbl7) [inline src/app/heroes/heroes.component.ts]: Consider removing this method

File: src/app/heroes/heroes.component.ts [hunk 9]
```diff
@@ -36,7 +39,8 @@ export class HeroesComponent implements OnInit {
 
   initHeroes() {
     this.getHeroes()
-      .concat(this.search())
+      .merge(this.search())
+      .merge(this.page())
```
Comment (dbl7) [inline src/app/heroes/heroes.component.ts]: Consider getting rid of 3 streams and merge, try using only two streams with concat

File: src/app/heroes/heroes.component.ts [hunk 10]
```diff
@@ -26,6 +28,7 @@ export class HeroesComponent implements OnInit {
   heroes: Hero[];
   totalHeroes: number;
   loading$ = new BehaviorSubject<boolean>(false);
+  pageNumber$ = new Subject<number>();
```
Comment (dbl7) [inline src/app/heroes/heroes.component.ts]: Use `BehaviorSubject` to initialize the first page

File: src/app/heroes/heroes.component.html [hunk 11]
```diff
@@ -29,7 +29,11 @@ <h1>Star Wars <small>Encyclopedia</small></h1><hr/>
             </tbody>
           </table>
           <app-spinner *ngIf="loading$ | async"></app-spinner>
-          <app-pagination *ngIf="!(loading$ | async)" [totalHeroes]="totalHeroes"></app-pagination>
+          <app-pagination 
+            *ngIf="!(loading$ | async)" 
+            [totalHeroes]="totalHeroes"
+            (pageChanged)="pageChanged($event)"
```
Comment (dbl7) [inline src/app/heroes/heroes.component.html]: Consider getting rid of this callback

**Review by dbl7 — CHANGES_REQUESTED** [review]
Great job! Try improving your implementation according with my comments.

File: src/app/heroes/heroes.component.ts (hunk 10 above)
Comment (dbl7) [reply src/app/heroes/heroes.component.ts]: Rename to `currentPage$`

**Review by dbl7 — CHANGES_REQUESTED** [review]

File: src/app/heroes/heroes.component.html (hunk 6 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.html]: Fixed

File: src/app/heroes/pagination/pagination.component.ts (hunk 7 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.ts]: Done

File: src/app/heroes/heroes.component.ts (hunk 8 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.ts]: Done

File: src/app/heroes/heroes.component.html (hunk 11 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.html]: Done

File: src/app/heroes/heroes.component.ts (hunk 10 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.ts]: Done

File: src/app/heroes/heroes.component.ts (hunk 9 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.ts]: Done

File: src/app/heroes/heroes.component.ts [hunk 12]
```diff
@@ -39,8 +39,7 @@ export class HeroesComponent implements OnInit {
 
   initHeroes() {
     this.getHeroes()
-      .merge(this.search())
-      .merge(this.page())
+      .concat(this.search())
```
Comment (dbl7) [inline src/app/heroes/heroes.component.ts]: I believe that `concat` doesn't work here, please use `merge`

File: src/app/heroes/heroes.component.ts [hunk 13]
```diff
@@ -49,8 +48,10 @@ export class HeroesComponent implements OnInit {
 
   getHeroes() {
     this.loading$.next(true);
```
Comment (dbl7) [inline src/app/heroes/heroes.component.ts]: I believe you turn on spinner two lines below

**Review by dbl7 — CHANGES_REQUESTED** [review]
Split getHeroes method in two getHeroes and search methods

Hide pagination until heroes are received first time

File: src/app/heroes/heroes.component.ts (hunk 12 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.ts]: Fixed

**Review by dbl7 — APPROVED** [review]

File: src/app/heroes/heroes.component.ts (hunk 13 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.component.ts]: Fixed

**Review by dbl7 — CHANGES_REQUESTED** [review]
Split getHeroes method in two getHeroes and search methods

Hide pagination until heroes are received first time

File: src/app/heroes/pagination/pagination.component.html [hunk 14]
```diff
@@ -9,6 +9,7 @@
     <li 
       class="page-item"
       *ngFor="let pageNumber of pageList$ | async"
+      [ngClass]="{'active': currentPage$.value === pageNumber}"
```
Comment (dbl7) [inline src/app/heroes/pagination/pagination.component.html]: Please, use shorthand version `[class.active]="condition"`

File: src/app/heroes/heroes.service.ts [hunk 15]
```diff
@@ -1,17 +1,34 @@
...
+      this.params = this.params.set('search', searchTerm);
+    }
+    
+    this.params = this.params.set('page', pageNumber);
+    
+    return this.http.get(`https://swapi.co/api/people/`, { params: this.params } )
+      .map((data: any) => ({
+        count: data.count,
+        heroes: data.results
+      }));
```
Comment (dbl7) [inline src/app/heroes/heroes.service.ts]: I believe that you can get rid of duplication of two map methods.

File: src/app/heroes/heroes.service.ts [hunk 16]
```diff
@@ -1,17 +1,34 @@
...
+    if (searchTerm) {
+      this.params = this.params.set('search', searchTerm);
+    }
+    
+    this.params = this.params.set('page', pageNumber);
+    
+    return this.http.get(`https://swapi.co/api/people/`, { params: this.params } )
+      .map((data: any) => ({
+        count: data.count,
+        heroes: data.results
```
Comment (dbl7) [inline src/app/heroes/heroes.service.ts]: I think you guess yourself what's wrong here

File: src/app/heroes/heroes.service.ts [hunk 17]
```diff
@@ -1,17 +1,34 @@
...
-import { HttpClient } from '@angular/common/http';
+import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
 import { Observable } from 'rxjs/Observable';
 import 'rxjs/add/operator/map';
 
 import { Hero } from '../models/hero';
 
 @Injectable()
 export class HeroesService {
+  private params: HttpParams = new HttpParams();
```
Comment (dbl7) [inline src/app/heroes/heroes.service.ts]: Why do you use global property for entire class? Just use params variable in methods where it's needed.

File: src/app/heroes/heroes.service.ts [hunk 18]
```diff
@@ -1,17 +1,34 @@
...
+      .map((data: any) => ({
+        count: data.count,
+        heroes: data.results
+      }));
+  }
+
+  search(searchTerm) {
+    this.params = this.params.set('search', searchTerm);
+
+    return this.http.get(`https://swapi.co/api/people/`, { params: this.params })
```
Comment (dbl7) [inline src/app/heroes/heroes.service.ts]: Don't add `/` at the end of the url, your url will look like `https://swapi.co/api/people/?page=1` and it's a little bit strange.

**Review by dbl7 — CHANGES_REQUESTED** [review]
I think something went wrong with your pagination

image

File: src/app/heroes/heroes.service.ts (hunk 18 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.service.ts]: Fixed

File: src/app/heroes/pagination/pagination.component.html (hunk 14 above)
Comment (dhutaryan) [reply src/app/heroes/pagination/pagination.component.html]: Done

File: src/app/heroes/heroes.service.ts (hunk 17 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.service.ts]: Fixed

File: src/app/heroes/heroes.service.ts (hunk 16 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.service.ts]: Done

File: src/app/heroes/heroes.service.ts (hunk 15 above)
Comment (dhutaryan) [reply src/app/heroes/heroes.service.ts]: Done

**Review by dbl7 — APPROVED** [review]
Great job!👍🏾

**Pull request merged**